hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/ContainerStateMachine.java

```diff
@@ -648,25 +651,38 @@ private ByteString readStateMachineData(
       // readStateMachineData should only be called for "write" to Ratis.
       Preconditions.checkArgument(!HddsUtils.isReadOnly(requestProto));
       if (requestProto.getCmdType() == Type.WriteChunk) {
-        final CompletableFuture<ByteString> future = new CompletableFuture<>();
-        ByteString data = stateMachineDataCache.get(entry.getIndex());
+        final CompletableFuture<ByteString> readStateMachineFuture =
+            new CompletableFuture<>();
+        long index = entry.getIndex();
+        ByteString data = stateMachineDataCache.get(index);
         if (data != null) {
-          future.complete(data);
-          return future;
+          readStateMachineFuture.complete(data);
+          return readStateMachineFuture;
         }
 
-        CompletableFuture.supplyAsync(() -> {
-          try {
-            future.complete(
-                readStateMachineData(requestProto, entry.getTerm(),
-                    entry.getIndex()));
-          } catch (IOException e) {
-            metrics.incNumReadStateMachineFails();
-            future.completeExceptionally(e);
-          }
-          return future;
-        }, getChunkExecutor(requestProto.getWriteChunk()));
-        return future;
+        // Queue the readStateMachine data readStateMachineFuture behind the
+        // write chunk for the same log index if it's under progress.
+        // Otherwise, issue the read
+        // chunk call over any chunk executor thread. Idea here is to ensure,
+        // read chunk always starts once the write chunk for the same log index
+        // finishes.
+        CompletableFuture<Message> response =
+            writeChunkFutureMap.computeIfPresent(index, (key, val) -> {
```

Comment:
computeIfPresent(..) will put a read future to the writeChunkFutureMap, which should only contain write futures.  We may change the code as below
```
      if (requestProto.getCmdType() == Type.WriteChunk) {
        final long index = entry.getIndex();
        final ByteString data = stateMachineDataCache.get(index);
        if (data != null) {
          return CompletableFuture.completedFuture(data);
        }

        final CompletableFuture<Message> writeFuture
            = Optional.ofNullable(writeChunkFutureMap.get(index))
            .orElse(CompletableFuture.completedFuture(null));

        // Queue the readStateMachine data readStateMachineFuture behind the
        // write chunk for the same log index if it's under progress.
        // Otherwise, issue the read
        // chunk call over any chunk executor thread. Idea here is to ensure,
        // read chunk always starts once the write chunk for the same log index
        // finishes.
        return writeFuture.thenCompose(
            w -> getReadStateMachineFuture(requestProto, entry));
      } else {
```